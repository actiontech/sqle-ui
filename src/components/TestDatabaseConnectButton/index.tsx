import { CheckCircleFilled, CloseCircleFilled } from '@ant-design/icons';
import { useBoolean } from 'ahooks';
import { Button, Space, Tooltip, Typography } from 'antd';
import { useTranslation } from 'react-i18next';
import EmptyBox from '../EmptyBox';
import { TestDatabaseConnectButtonProps } from './index.type';

const TestDatabaseConnectButton: React.FC<TestDatabaseConnectButtonProps> = (
  props
) => {
  const { t } = useTranslation();
  const [initHide, { setFalse }] = useBoolean(props.initHide ?? true);

  const handleClick = () => {
    if (initHide && props.initHide === undefined) {
      setFalse();
    }
    props.onClickTestButton();
  };

  return (
    <Space>
      <Button onClick={handleClick} loading={props.loading}>
        {t('dataSource.dataSourceForm.testDatabaseConnection')}
      </Button>
      <EmptyBox if={props.initHide !== undefined ? !props.initHide : !initHide}>
        {props.loading && (
          <Typography.Link>
            {t('dataSource.dataSourceForm.testing')}
          </Typography.Link>
        )}
        {!props.loading && props.connectAble && (
          <Typography.Text type="success">
            <CheckCircleFilled />
            {t('dataSource.dataSourceForm.testSuccess')}
          </Typography.Text>
        )}
        {!props.loading && !props.connectAble && (
          <Tooltip overlay={props.connectDisableReason}>
            <Typography.Text type="danger">
              <CloseCircleFilled />
              {t('dataSource.dataSourceForm.testFailed')}
            </Typography.Text>
          </Tooltip>
        )}
      </EmptyBox>
    </Space>
  );
};

export default TestDatabaseConnectButton;
