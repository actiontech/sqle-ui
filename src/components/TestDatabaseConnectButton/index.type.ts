export type TestDatabaseConnectButtonProps = {
  initHide?: boolean;
  onClickTestButton: () => void;
  loading: boolean;
  connectAble: boolean;
  connectDisableReason?: string;
};
