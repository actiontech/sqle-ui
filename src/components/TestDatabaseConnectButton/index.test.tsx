import TestDatabaseConnectButton from '.';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';

describe('TestDatabaseConnectButton', () => {
  const buttonText = 'dataSource.dataSourceForm.testDatabaseConnection';

  describe('uncontrolled mode', () => {
    test('should render message by initHide', () => {
      const { container, rerender } = render(
        <TestDatabaseConnectButton
          onClickTestButton={jest.fn}
          loading={true}
          connectAble={true}
        />
      );
      expect(container).toMatchSnapshot();

      rerender(
        <TestDatabaseConnectButton
          onClickTestButton={jest.fn}
          loading={false}
          connectAble={true}
        />
      );

      expect(container).toMatchSnapshot();
    });

    test('should show connect result when click the test button', () => {
      const handleClickButton = jest.fn();
      const { container, rerender } = render(
        <TestDatabaseConnectButton
          onClickTestButton={handleClickButton}
          loading={true}
          connectAble={true}
        />
      );
      fireEvent.click(screen.getByText(buttonText));
      expect(handleClickButton).not.toBeCalled();

      rerender(
        <TestDatabaseConnectButton
          onClickTestButton={handleClickButton}
          loading={false}
          connectAble={true}
        />
      );
      expect(container).toMatchSnapshot();
      fireEvent.click(screen.getByText(buttonText));
      expect(handleClickButton).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();

      handleClickButton.mockClear();
      cleanup();

      const { container: disableContainer } = render(
        <TestDatabaseConnectButton
          onClickTestButton={handleClickButton}
          loading={false}
          connectAble={false}
          connectDisableReason="disable reason"
        />
      );
      expect(disableContainer).toMatchSnapshot();
      fireEvent.click(screen.getByText(buttonText));
      expect(handleClickButton).toBeCalledTimes(1);
      expect(disableContainer).toMatchSnapshot();
    });
  });

  describe('controlled mode', () => {
    test('should render message by initHide', () => {
      const { container, rerender } = render(
        <TestDatabaseConnectButton
          onClickTestButton={jest.fn}
          initHide={true}
          loading={false}
          connectAble={false}
        />
      );
      expect(container).toMatchSnapshot();

      rerender(
        <TestDatabaseConnectButton
          onClickTestButton={jest.fn}
          initHide={true}
          loading={true}
          connectAble={false}
        />
      );

      expect(container).toMatchSnapshot();

      rerender(
        <TestDatabaseConnectButton
          onClickTestButton={jest.fn}
          initHide={false}
          loading={true}
          connectAble={false}
          connectDisableReason="disable reason"
        />
      );

      expect(container).toMatchSnapshot();

      rerender(
        <TestDatabaseConnectButton
          onClickTestButton={jest.fn}
          initHide={false}
          loading={false}
          connectAble={true}
          connectDisableReason="disable reason"
        />
      );

      expect(container).toMatchSnapshot();

      rerender(
        <TestDatabaseConnectButton
          onClickTestButton={jest.fn}
          initHide={false}
          loading={false}
          connectAble={false}
          connectDisableReason="disable reason"
        />
      );

      expect(container).toMatchSnapshot();
    });

    test('should call onClickTestButton when click the test button', () => {
      const handleClickTestButton = jest.fn();
      const { container } = render(
        <TestDatabaseConnectButton
          onClickTestButton={handleClickTestButton}
          initHide={true}
          loading={false}
          connectAble={false}
          connectDisableReason="disable reason"
        />
      );
      expect(container).toMatchSnapshot();
      fireEvent.click(screen.getByText(buttonText));
      expect(handleClickTestButton).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();
    });
  });
});
