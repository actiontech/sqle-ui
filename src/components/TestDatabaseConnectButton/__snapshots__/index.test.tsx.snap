// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestDatabaseConnectButton controlled mode should call onClickTestButton when click the test button 1`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    />
  </div>
</div>
`;

exports[`TestDatabaseConnectButton controlled mode should call onClickTestButton when click the test button 2`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        ant-click-animating-without-extra-node="false"
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    />
  </div>
</div>
`;

exports[`TestDatabaseConnectButton controlled mode should render message by initHide 1`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    />
  </div>
</div>
`;

exports[`TestDatabaseConnectButton controlled mode should render message by initHide 2`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        class="ant-btn ant-btn-loading"
        type="button"
      >
        <span
          class="ant-btn-loading-icon"
        >
          <span
            aria-label="loading"
            class="anticon anticon-loading anticon-spin"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="loading"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
              />
            </svg>
          </span>
        </span>
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    />
  </div>
</div>
`;

exports[`TestDatabaseConnectButton controlled mode should render message by initHide 3`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        class="ant-btn ant-btn-loading"
        type="button"
      >
        <span
          class="ant-btn-loading-icon"
        >
          <span
            aria-label="loading"
            class="anticon anticon-loading anticon-spin"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="loading"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
              />
            </svg>
          </span>
        </span>
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    >
      <a
        class="ant-typography"
      >
        dataSource.dataSourceForm.testing
      </a>
    </div>
  </div>
</div>
`;

exports[`TestDatabaseConnectButton controlled mode should render message by initHide 4`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    >
      <span
        class="ant-typography ant-typography-success"
      >
        <span
          aria-label="check-circle"
          class="anticon anticon-check-circle"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="check-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm193.5 301.7l-210.6 292a31.8 31.8 0 01-51.7 0L318.5 484.9c-3.8-5.3 0-12.7 6.5-12.7h46.9c10.2 0 19.9 4.9 25.9 13.3l71.2 98.8 157.2-218c6-8.3 15.6-13.3 25.9-13.3H699c6.5 0 10.3 7.4 6.5 12.7z"
            />
          </svg>
        </span>
        dataSource.dataSourceForm.testSuccess
      </span>
    </div>
  </div>
</div>
`;

exports[`TestDatabaseConnectButton controlled mode should render message by initHide 5`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    >
      <span
        class="ant-typography ant-typography-danger"
      >
        <span
          aria-label="close-circle"
          class="anticon anticon-close-circle"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
            />
          </svg>
        </span>
        dataSource.dataSourceForm.testFailed
      </span>
    </div>
  </div>
</div>
`;

exports[`TestDatabaseConnectButton controlled mode should set initHide to false when click the button 1`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    />
  </div>
</div>
`;

exports[`TestDatabaseConnectButton controlled mode should set initHide to false when click the button 2`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        ant-click-animating-without-extra-node="false"
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    >
      <span
        class="ant-typography ant-typography-danger"
      >
        <span
          aria-label="close-circle"
          class="anticon anticon-close-circle"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
            />
          </svg>
        </span>
        dataSource.dataSourceForm.testFailed
      </span>
    </div>
  </div>
</div>
`;

exports[`TestDatabaseConnectButton uncontrolled mode should render message by initHide 1`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        class="ant-btn ant-btn-loading"
        type="button"
      >
        <span
          class="ant-btn-loading-icon"
        >
          <span
            aria-label="loading"
            class="anticon anticon-loading anticon-spin"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="loading"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
              />
            </svg>
          </span>
        </span>
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    />
  </div>
</div>
`;

exports[`TestDatabaseConnectButton uncontrolled mode should render message by initHide 2`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    />
  </div>
</div>
`;

exports[`TestDatabaseConnectButton uncontrolled mode should show connect result when click the test button 1`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        ant-click-animating-without-extra-node="false"
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    />
  </div>
</div>
`;

exports[`TestDatabaseConnectButton uncontrolled mode should show connect result when click the test button 2`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        ant-click-animating-without-extra-node="false"
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    >
      <span
        class="ant-typography ant-typography-success"
      >
        <span
          aria-label="check-circle"
          class="anticon anticon-check-circle"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="check-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm193.5 301.7l-210.6 292a31.8 31.8 0 01-51.7 0L318.5 484.9c-3.8-5.3 0-12.7 6.5-12.7h46.9c10.2 0 19.9 4.9 25.9 13.3l71.2 98.8 157.2-218c6-8.3 15.6-13.3 25.9-13.3H699c6.5 0 10.3 7.4 6.5 12.7z"
            />
          </svg>
        </span>
        dataSource.dataSourceForm.testSuccess
      </span>
    </div>
  </div>
</div>
`;

exports[`TestDatabaseConnectButton uncontrolled mode should show connect result when click the test button 3`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    />
  </div>
</div>
`;

exports[`TestDatabaseConnectButton uncontrolled mode should show connect result when click the test button 4`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <button
        ant-click-animating-without-extra-node="false"
        class="ant-btn"
        type="button"
      >
        <span>
          dataSource.dataSourceForm.testDatabaseConnection
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    >
      <span
        class="ant-typography ant-typography-danger"
      >
        <span
          aria-label="close-circle"
          class="anticon anticon-close-circle"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
            />
          </svg>
        </span>
        dataSource.dataSourceForm.testFailed
      </span>
    </div>
  </div>
</div>
`;
